src/audio/base_fw.c

```diff
@@ -234,24 +236,46 @@ static int basefw_mem_state_info(uint32_t *data_offset, char *data)
 	return 0;
 }
 
+static log_timestamp_t basefw_get_timestamp(void)
+{
+	return sof_cycle_get_64() + time_delta;
+}
+
 static uint32_t basefw_set_system_time(uint32_t param_id,
 				       bool first_block,
 				       bool last_block,
 				       uint32_t data_offset,
 				       const char *data)
 {
-	/* TODO: configurate time to logging subsystem */
+	uint64_t host_time, host_time_cycle;
+	uint64_t current_dsp_time;
+	uint64_t current_dsp_cycle;
+
 	if (!(first_block && last_block))
 		return IPC4_INVALID_REQUEST;
 
 	global_system_time_info.host_time.val_l = ((const struct ipc4_system_time *)data)->val_l;
 	global_system_time_info.host_time.val_u = ((const struct ipc4_system_time *)data)->val_u;
 
-	uint64_t current_dsp_time = sof_cycle_get_64();
+	current_dsp_cycle = sof_cycle_get_64();
+	current_dsp_time = k_cyc_to_us_ceil64(current_dsp_cycle);
 
 	global_system_time_info.dsp_time.val_l = (uint32_t)(current_dsp_time);
 	global_system_time_info.dsp_time.val_u = (uint32_t)(current_dsp_time >> 32);
 
+	/* use default timestamp if 64bit is not enabled since 64bit is necessary for host time */
+	if (!IS_ENABLED(CONFIG_LOG_TIMESTAMP_64BIT)) {
+		LOG_WRN("64bits timestamp is disabled, so use default timestamp");
+		return IPC4_SUCCESS;
+	}
+
+	host_time = global_system_time_info.host_time.val_l |
+		((uint64_t)global_system_time_info.host_time.val_u << 32);
```

Comment:
Why not just read this directly from `*data` as you were earlier?

Reply:
you commented it was weird so I made this more clear

Reply:
What's weird is the same `*data` goes  in the same function to two different places that seem to do more or less the same thing without a word of explanation of why these two places are needed and how they compare to each other. That has never changed in this PR and it has never been explained.